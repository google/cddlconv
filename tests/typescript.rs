use std::io::BufWriter;

use cddl::visitor::Visitor;
use cddlconv;

#[test]
fn it_works() {
    let input = std::fs::read_to_string("examples/webdriver-bidi/webdriver-bidi.cddl").unwrap();
    let cddl = cddl::parser::cddl_from_str(&input, true).unwrap();
    let stdout = BufWriter::new(Vec::new());
    let stderr = BufWriter::new(Vec::new());
    let mut engine = cddlconv::engines::typescript::Engine::with_writers(stdout, stderr);
    engine.visit_cddl(&cddl).unwrap();
    engine.print_postamble();

    let (stdout, stderr) = engine.into_writers();
    insta::assert_snapshot!(String::from_utf8(stderr.into_inner().unwrap()).unwrap());
    insta::assert_snapshot!(String::from_utf8(stdout.into_inner().unwrap()).unwrap());
}

#[test]
fn it_works_with_arrays() {
    let input = std::fs::read_to_string("examples/rfc-examples/arrays.cddl").unwrap();
    let cddl = cddl::parser::cddl_from_str(&input, true).unwrap();
    let stdout = BufWriter::new(Vec::new());
    let stderr = BufWriter::new(Vec::new());
    let mut engine = cddlconv::engines::typescript::Engine::with_writers(stdout, stderr);
    engine.visit_cddl(&cddl).unwrap();
    engine.print_postamble();

    let (stdout, stderr) = engine.into_writers();
    insta::assert_snapshot!(String::from_utf8(stderr.into_inner().unwrap()).unwrap());
    insta::assert_snapshot!(String::from_utf8(stdout.into_inner().unwrap()).unwrap());
}

#[test]
fn it_works_with_maps() {
    let input = std::fs::read_to_string("examples/rfc-examples/maps.cddl").unwrap();
    let cddl = cddl::parser::cddl_from_str(&input, true).unwrap();
    let stdout = BufWriter::new(Vec::new());
    let stderr = BufWriter::new(Vec::new());
    let mut engine = cddlconv::engines::typescript::Engine::with_writers(stdout, stderr);
    engine.visit_cddl(&cddl).unwrap();
    engine.print_postamble();

    let (stdout, stderr) = engine.into_writers();
    insta::assert_snapshot!(String::from_utf8(stderr.into_inner().unwrap()).unwrap());
    insta::assert_snapshot!(String::from_utf8(stdout.into_inner().unwrap()).unwrap());
}
