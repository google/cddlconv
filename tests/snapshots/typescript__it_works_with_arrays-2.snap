---
source: tests/typescript.rs
assertion_line: 33
expression: "String::from_utf8(stdout.into_inner().unwrap()).unwrap()"
---
export type EquipmentType = ([
(string),(string)]);
export type EquipmentType2 = ([

/**
 * @defaultValue `"hello"`
 */
(string),(string)]);
export type EquipmentType3 = ([
...([]|[(string),]),(string)]);
export type EquipmentTolerances = ([
...(([
(number),(number)])[])]);
export type Person = ({
  "name":(string),  "age":(number)});
export type UnlimitedPeople = ([
...((Person)[])]);
export type OneOrTwoPeople = ([
...(|[(Person),]|[(Person),(Person),])]);
export type AtMostTwoPeople = ([
...([]|[(Person),]|[(Person),(Person),])]);
export type AtLeastTwoPeople = ([
...((Person)[])]);
export type OnePerson = ([
(Person)]);
export type TwoPerson = ([
(Person),(Person)]);
export type OptionalPerson = ([
...([]|[(Person),])]);
export type PersonWithEquipment = ([
...((Person)[]),...((EquipmentType)[])]);
export type PersonWithAttrNames = ([
...((Person)[])]);
export type PersonWithAttrNames1 = ([
...((Person)[]),...((EquipmentType)[])]);
export type PersonWithAttrNames2 = ([
...((Person)[]),...((EquipmentType)[])]);
export type PersonWithAttrNames3 = ([
...((Person)[]),...([]|[(EquipmentType),])]);

